### PR TITLE
Added Edge Detection module

### PR DESCRIPTION
**Pull request by tech4GT**
image1_0

image1_1

@jywarren please have a look , right now exporting the edge detected image, will create a separate issue for adding support for json output, what say?

### CODE REVIEW

**Comment by jywarren** [comment]
Oh, this is cool! Could you try running it on the image here and writing back in that thread? I'd like to get some input from these folks:

https://publiclab.org/questions/jlev/02-01-2018/how-can-i-identify-bits-of-plastic-from-the-beach-in-an-image

This is so cool! @ccpandhare what do you think?

**Comment by tech4GT** [comment]
image1_1
@jywarren there is one limitation though, this only works with png images right now, but i will add support

**Comment by tech4GT** [comment]
@jywarren right now this is basic edge detection we can use algorithms like hysteresis to improve the quality of edge detection😁

**Comment by tech4GT** [comment]
@jywarren since jpg does not support transparency , can i convert an input jpg file to png and convert the output back to jpeg, will that be an acceptable solution?

**Comment by ccpandhare** [comment]
This is so cool!!!

@jywarren One more off the bucket list. :-)

**Comment by ccpandhare** [comment]
Why isn't this working with JPEG?

**Comment by jywarren** [comment]
Oh this is cool! I echo chinmay's question!

Please post to the plastics question, they'll be interested to see this!!!

**Comment by tech4GT** [comment]
@jywarren @ccpandhare  actually jpegs do not support transparency values and edge detection requires that
a values are 255 for each pixel in jpeg

**Comment by tech4GT** [comment]
@jywarren should i post the complete code on the plastics question?

**Comment by jywarren** [comment]
Cool! I think you could just post the image of edges and link back here.

**Comment by tech4GT** [comment]
@jywarren done😁

**Comment by tech4GT** [comment]
@jywarren right now i have just implemented Gaussian blur with sobel filter on the image after conversion to greyscale hence basic edge detection, i will implement remaining steps for canny edge detection in future which will finally output image  with black edges surrounding white regions

**Comment by tech4GT** [comment]
@jywarren right now this outputs a greyscale image you are right😁

**Comment by jywarren** [comment]
That's awesome. Im thinking some of this explanation could go into the
module description to help people understand what it does and what the
output means. What do you think?

**Comment by tech4GT** [comment]
@jywarren sure no problem but right now I don't have access to a computer I'll do it a bit later if that's fine?😅

**Comment by tech4GT** [comment]
@jywarren please have a look if anything else needs to be added, rest assured this is ready to be merged and i will surely open issues for enhancing this module and adding tests, thanks a lot😁

**Comment by tech4GT** [comment]
@jywarren is there anything else i can do in this😅

File: src/modules/EdgeDetect/info.json [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
+    "name": "Detect Edges",
+    "description": "Detects the edges in an image and outputs a a greyscale image with the canny method",
```
Comment (ccpandhare) [inline src/modules/EdgeDetect/info.json]: Please fix the double "a" in the description field's value

Comment (ccpandhare) [reply src/modules/EdgeDetect/info.json]: Again, I apologize for being that picky!

**Comment by ccpandhare** [comment]
Also, how would the edges thing work? Would it be like thresholding the greyscale output image of this module or would it have an entirely different approach?

**Comment by tech4GT** [comment]
@ccpandhare i was thinking once i implement complete canny edge detection we can get the edge pixels and output the positions, what say?

File: src/modules/EdgeDetect/info.json (hunk 1 above)
Comment (tech4GT) [reply src/modules/EdgeDetect/info.json]: @ccpandhare no apologies needed in fact thank you for being so observant😁

**Comment by jywarren** [comment]
Hi, this is looking really good; i wasn't previously familiar with how edge detection works, so I had to do some research to understand fully. 

With that in mind, I think we might add this to the description just for clarity:

```
Edge detection: this module detects edges using the Canny method, which first Gaussian blurs the image to reduce noise (amount of blur configurable in settings as `options.blur`), then applies a number of steps to highlight edges, resulting in a greyscale image where the brighter the pixel, the stronger the detected edge. Read more at: https://en.wikipedia.org/wiki/Canny_edge_detector
```

Do we also allow setting the threshold? https://en.wikipedia.org/wiki/Canny_edge_detector#Parameters

Finally, the Wikipedia page shows a black background instead of alpha. Would it be possible to do this too? Then it'd work in JPG too. 

This is really impressive work! Potentially we could base a Blur module off of components of this too!

**Comment by tech4GT** [comment]
@jywarren thanks a lot and yes we already have everything for the Gaussian blur so we can create a module for that.
secondly i am yet to implement all the steps of canny edge detection and hysteresis is one of them which takes the threshold values, hence i called this basic edge detection and i will create separate issue for adding the rest of the steps and i will shortly add them all
Thirdly the black background thing can be implemented after the rest of the steps are implemented which i am already working on and i will submit the PR in a week or so, i hope that's fine.

**Comment by tech4GT** [comment]
@jywarren i have also taken care of the changes that @ccpandhare mentioned😁

**Comment by tech4GT** [comment]
@jywarren finally done with this, this was a tough one😅. Full canny edge detection🎉

**Comment by tech4GT** [comment]
image1_1

**Comment by tech4GT** [comment]
@jywarren now this also takes low and high threshold and we also have a edge pixel positions array which can be exported as json as well

**Comment by jywarren** [comment]
Hi, this sounds great and looks great. If you can make a follow-up issue to make the background black, and for any other follow-up issues, then this sounds good and I'm merging it in! Thanks!

**Comment by jywarren** [comment]
I just wanted to say that this was a very impressive project! 🎉 🚀 👍

**Comment by tech4GT** [comment]
@jywarren thanks a lot😁

**Pull request merged**